File: 003_longest_substring_without_repeating_characters/Solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set = set()
        rk, ans = -1, 0
        length = len(s)

        for i in range(length):
            if i != 0: char_set.remove(s[i - 1])

            while rk + 1 < length and s[rk + 1] not in char_set:
                char_set.add(s[rk + 1])
                rk += 1
            ans = max(ans, rk - i + 1)

        return ans


if __name__ == '__main__':
    solution = Solution()
    print(solution.lengthOfLongestSubstring('abcabcbb'))

File: 003_longest_substring_without_repeating_characters/Solution.java
import java.util.HashSet;

public class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> charSet = new HashSet<>();
        int length = s.length();
        int rk = -1, ans = 0;

        for (int i = 0; i < length; i++) {
            if (i != 0) {
                charSet.remove(s.charAt(i - 1));
            }
            while (rk + 1 < length && !charSet.contains(s.charAt(rk + 1))) {
                charSet.add(s.charAt(rk + 1));
                rk++;
            }
            ans = Math.max(ans, rk - i + 1);
        }

        return ans;
    }

    public static void main(String[] args) {
        String s = "abcabcbb";
        Solution solution = new Solution();
        System.out.println(solution.lengthOfLongestSubstring(s));
    }
}
